xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,251 @@
+
+            @Override
+            public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+              if (clusterWatcher == null) {
+                clusterWatcher = new ClusterWatcherImpl(helper, resolvedAddresses);
+                xdsClient.watchClusterData(newCdsConfig.name, clusterWatcher);
+                if (oldCdsConfig != null) {
+                  xdsClient.cancelClusterDataWatch(oldCdsConfig.name, oldClusterWatcher);
+                }
+                CdsLoadBalancer.this.clusterWatcher = clusterWatcher;
+              }
+            }
+          };
+        }
+      };
+
+      switchingLoadBalancer.switchTo(fixedCdsConfigBalancer);
+    }
+
+    switchingLoadBalancer.handleResolvedAddresses(resolvedAddresses);
+
+    // clusterWatcher is also updated after switchingLoadBalancer.handleResolvedAddresses
+    cdsConfig = newCdsConfig;
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    channelLogger.log(ChannelLogLevel.DEBUG, "Name resolution error: '%s'", error);
+  }
+
+  @Override
+  public boolean canHandleEmptyAddressListFromNameResolution() {
+    return true;
+  }
+
+  @Override
+  public void shutdown() {
+    channelLogger.log(ChannelLogLevel.DEBUG, "CDS load balancer is shutting down");
+
+    switchingLoadBalancer.shutdown();
+    if (xdsClientRef != null) {
+      xdsClientRef.returnObject(xdsClient);
+    }
+  }
+
+  private final class ClusterWatcherImpl implements ClusterWatcher {
+
+    final LoadBalancer edsBalancer;
+    final Helper helper;
+    final ResolvedAddresses resolvedAddresses;
+
+    ClusterWatcherImpl(Helper helper, ResolvedAddresses resolvedAddresses) {
+      this.helper = helper;
+      this.resolvedAddresses = resolvedAddresses;
+      this.edsBalancer = lbRegistry.getProvider("xds_experimental").newLoadBalancer(helper);
+    }
+
+    @Override
+    public void onClusterChanged(ClusterUpdate update) {
+      Preconditions.checkArgument(
+          update.getLbPolicy().equals("round_robin"),
+          "The lbPolicy in ClusterUpdate '%s' is not 'round_robin'", update);
+
+      final XdsConfig edsConfig = new XdsConfig(
+          /* balancerName = */ null,
+          new LbConfig(update.getLbPolicy(), ImmutableMap.<String, Object>of()),
+          /* fallbackPolicy = */ null,
+          /* edsServiceName = */ update.getEdsServiceName(),
+          /* lrsServerName = */ update.getLrsServerName());
+
+      LoadStatsStore loadStatsStore = new LoadStatsStoreImpl();
+      if (update.isEnableLrs()) {
+        xdsClient.reportClientStats(
+            update.getClusterName(), update.getLrsServerName(), loadStatsStore);
+      }
+      edsBalancer.handleResolvedAddresses(
+          resolvedAddresses.toBuilder()
+              .setAttributes(
+                  resolvedAddresses.getAttributes().toBuilder()
+                      .discard(ATTR_LOAD_BALANCING_CONFIG)
+                      .set(XdsAttributes.LOAD_STATS_STORE_REF, loadStatsStore)
```

Comment:
I am trying to understand what happens when `update.isEnableLrs()` is `false`. How about doing this:
```
      LoadStatsStore loadStatsStore = null;
      if (update.isEnableLrs()) {
        loadStatsStore = new LoadStatsStoreImpl();
        xdsClient.reportClientStats(
            update.getClusterName(), update.getLrsServerName(), loadStatsStore);
      }
      edsBalancer.handleResolvedAddresses(
          resolvedAddresses.toBuilder()
              .setAttributes(
                  resolvedAddresses.getAttributes().toBuilder()
                      .discard(ATTR_LOAD_BALANCING_CONFIG)
                      .set(XdsAttributes.LOAD_STATS_STORE_REF, loadStatsStore)
                      .build())
              .setLoadBalancingPolicyConfig(edsConfig)
          .build());
    }
```

Otherwise you are passing the loadStatsStore to individual subchannels but will never be used for that cluster.

Reply:
LoadStatsStore is always needed, the problem is we'd be better to cache loadStatsStore for each cluster name, so that it could be reused once lrs is enabled from being disabled. + @voidzcy ?

Reply:
Consider the case when lrs is changed from disabled to enabled. With the current flow, you'll just create a new loadStatsStore instance and set it in xdsClient for the cluster-name and pass it to the attribute. So the old instance was completely overwritten and data from it lost?

Reply:
Yeah, we should only create a new `LoadStatsStore` instance when we switch cluster.

Reply:
Now cached the stats store for each cluster.